### PR TITLE
Enhance DemoteLocalTemporariesToVariablesPass

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Description

This PR is intended to expand the number of temporary fields that can be reduced to scalars by the `DemoteLocalTemporariesToVariablesPass`. The existing implementation will not demote any temporary that is referenced across more than one stage. However, a temporary field can also be demoted to a scalar if it is always written before being read in each stage where it appears. Given the following example:
```python
@gtscript.stencil
def set_vals(gam: FloatField, delp: FloatField):
    with computation(PARALLEL), interval(0, 1):
        grid_ratio = delp[0, 0, 1] / delp
        bet = grid_ratio * (grid_ratio + 0.5)
        gam = (1.0 + grid_ratio * (grid_ratio + 1.5)) / bet
    with computation(FORWARD), interval(1, 2):
        grid_ratio = delp[0, 0, -1] / delp
        bet = 2.0 + grid_ratio + grid_ratio - gam
        gam = (3.0 * (gam[0, 0, -1] + gam)) / bet
```

The temporary fields, `grid_ratio` and `bet` will not be demoted in the current implementation as they are accessed within multiple stages. However, since both are written before read in both stages they can be demoted since the values from the previous stages are overwritten.

This PR makes the following changes:
- The `DemoteLocalTemporariesToVariablesPass.CollectDemotableSymbols` inner class is refactored to capture the previously described visitor by implementing a `visit_Assign` method to distinguish between a `FieldRef` that is written (left hand) and those that are read (right hand).
- The `var_refs_defined` set in the `PythonSourceGenerator` class is cleared between stages. This prevents broadcast errors that can occur if a `VarRef` defined in a previous stage has a different shape and is passed as the `else_expr` to a `np.where` call when handling temporaries in conditionals.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these new features
- [x] All relevant documentation has been updated or added

### CODE REVIEW

File: src/gt4py/analysis/passes.py [hunk 1]
```diff
@@ -1194,38 +1194,37 @@ class DemoteLocalTemporariesToVariablesPass(TransformPass):
...
             return collector(node)
 
         def __call__(self, node: gt_ir.StencilImplementation) -> Set[str]:
             assert isinstance(node, gt_ir.StencilImplementation)
-            self.demotables = set(node.temporary_fields)
-            self.seen_symbols: Set[str] = set()
-            self.stage_symbols: Optional[Set[str]] = None
+            self.demotables: Dict[str, str] = {
+                field_name: "" for field_name in node.temporary_fields
+            }
```
Comment (jdahm) [inline src/gt4py/analysis/passes.py:1206]: Can you add a docstring below this that describes what `self.demotables` stores?

File: src/gt4py/analysis/passes.py [hunk 2]
```diff
@@ -1194,38 +1194,37 @@ class DemoteLocalTemporariesToVariablesPass(TransformPass):
...
             self.visit(node)
-            return self.demotables
+            return set(self.demotables.keys())
 
         def visit_Stage(self, node: gt_ir.Stage) -> None:
-            self.stage_symbols = set()
-
+            self.stage_name = node.name
             self.generic_visit(node)
```
Comment (jdahm) [inline src/gt4py/analysis/passes.py:1212]: I think `self.generic_visit` can pass `stage_name` down to `visit_Assign` by adding it as a keyword to the signature of `visit_Assign`. In that case `stage_name` does not need to be a member of the class.
```
def visit_Assign(self, node: gt_ir.Assign, *, stage_name: str) -> None:
    ...
```

File: src/gt4py/analysis/passes.py [hunk 3]
```diff
@@ -1194,38 +1194,37 @@ class DemoteLocalTemporariesToVariablesPass(TransformPass):
...
+            self.visit(node.value)
 
         def visit_FieldRef(self, node: gt_ir.FieldRef) -> None:
-            if any(v != 0 for v in node.offset.values()):
-                self.demotables.discard(node.name)
-            assert self.stage_symbols is not None
-            self.stage_symbols.add(node.name)
+            field_name = node.name
+            if field_name in self.demotables:
+                assert len(self.demotables[field_name]) > 0
```
Comment (jdahm) [inline src/gt4py/analysis/passes.py:1223]: Instead of asserting the `len(...) > 0`, the more Pythonic way is probably to make `self.demotables: Dict[str, Optional[str]]` and set it to `None` initially.

File: src/gt4py/backend/numpy_backend.py [hunk 4]
```diff
@@ -263,11 +263,6 @@ def _visit_branch_stmt(self, stmt: gt_ir.Statement) -> List[str]:
             target = self.visit(stmt.target)
             value = self.visit(stmt.value)
 
-            data_type = (
-                self.block_info.symbols[target].data_type
-                if target in self.block_info.symbols
-                else stmt.target.data_type
-            )
```
Comment (jdahm) [inline src/gt4py/backend/numpy_backend.py:270]: Is this a related change?

File: src/gt4py/backend/python_generator.py [hunk 5]
```diff
@@ -261,6 +261,7 @@ def visit_Stage(self, node: gt_ir.Stage, *, iteration_order):
         self.block_info.accessors = {accessor.symbol for accessor in node.accessors}
         self.block_info.iteration_order = iteration_order
         self.block_info.extent = node.compute_extent
+        self.var_refs_defined.clear()
```
Comment (jdahm) [inline src/gt4py/backend/python_generator.py:264]: Was this change intended?

**Review by jdahm — CHANGES_REQUESTED** [review]
Looks great, especially the tests! Just a few Pythonic style comments in `passes.py`.

File: src/gt4py/backend/numpy_backend.py (hunk 4 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/numpy_backend.py:270]: No, this was to satisfy the linter as `data_type` was unused.

File: src/gt4py/backend/python_generator.py (hunk 5 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/python_generator.py:264]: Yes, this change is necessary for conditionals to work properly in the `numpy` backend. Will add a test for that case.

File: src/gt4py/analysis/passes.py [hunk 6]
```diff
@@ -1201,27 +1201,27 @@ def apply(cls, node: gt_ir.StencilImplementation) -> Set[str]:
 
         def __call__(self, node: gt_ir.StencilImplementation) -> Set[str]:
             assert isinstance(node, gt_ir.StencilImplementation)
-            self.demotables: Dict[str, str] = {
-                field_name: "" for field_name in node.temporary_fields
+            """Dictionary mapping temporaries to their most recently referenced stage."""
+            self.demotables: Dict[str, Optional[str]] = {
+                temp_field: None for temp_field in node.temporary_fields
             }
```
Comment (jdahm) [inline src/gt4py/analysis/passes.py]: I believe docstrings should go after the variable definition. See https://www.python.org/dev/peps/pep-0224/#id4

File: tests/test_integration/stencil_definitions.py [hunk 7]
```diff
@@ -283,3 +283,17 @@ def allow_empty_computation(in_field: Field3D, out_field: Field3D):
...
+    in_field: gtscript.Field[float], out_field: gtscript.Field[float], m: int
+):
+    with computation(PARALLEL), interval(1, 2):
+        if m == 1:
+            tmp_field = in_field[0, 0, -1] / in_field
+            out_field = 2.0 * tmp_field
+    with computation(FORWARD), interval(2, None):
+        if m == 1:
+            tmp_field = 2.0 * in_field[0, 0, -1]
+            out_field = tmp_field - in_field
```
Comment (jdahm) [inline tests/test_integration/stencil_definitions.py:299]: Is there another test case that we could add this to? Every new test here adds quite a lot of test time.

**Review by jdahm — APPROVED** [review]
This looks great!

**Review by gronerl — APPROVED** [review]
Thanks, this looks good.
